### PR TITLE
Update format.ml to avoid Queue.empty

### PR DESCRIPTION
**Pull request by craff**
issue #13463

### CODE REVIEW

**Comment by gasche** [comment]
I would find the code easier to explain if we did *not* print the token if we realize that the queue has become empty, something like:

```
match take_opt pp_queue with
| None -> ()
| Some ->
  (* printing logic *)
```

In any case, I think there should be a comment that explains the choice to deal with concurrent races in best effort.

**Comment by gasche** [comment]
Note: I am not, a priori, convinced that we should try to slightly better support this use-case, instead of failing loudly with an exception (as we were doing). Why would you try to rely on this, instead of fixing your code to avoid racing on format buffers?

**Comment by craff** [comment]
Hello Gabriel,

The problem I had, is when you fall into the race condition while adding
debugging message to solve a problem and get an uncaught exception, you
may loose a lot of time. Undefined behavior and bad printing would have
be OK in this case, but instead I got a nightmarish debugging session!

I think uncaught exception is really bad in a printing function, that
can be a log inside an exception handler.

Gabriel Scherer ***@***.***> writes:

> I would find the code easier to explain if we did not print the token if we realize that the queue has become empty, something like:
>
> match take_opt pp_queue with
> | None -> ()
> | Some ->
>   (* printing logic *)
>
> In any case, I think there should be a comment that explains the choice to deal with concurrent races in best effort.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.


--
Christophe Raffalli
web: https://raffalli.eu

**Comment by craff** [comment]
> I would find the code easier to explain if we did _not_ print the token if we realize that the queue has become empty, something like:
> 
> ```
> match take_opt pp_queue with
> | None -> ()
> | Some ->
>   (* printing logic *)
> ```
> 
> In any case, I think there should be a comment that explains the choice to deal with concurrent races in best effort.

Done. Thanks for your comment.

**Comment by gasche** [comment]
Following the line of thoughts around "fail loudly if there is an issue": would failing with an `Invalid_argument` exception be okay, or are you saying that your code is catching all exceptions (including those that should not be recovered from) and that also would not work? If that would have worked for you, I think that `invalid_arg "Format: unsychronized concurrent access"` would have my preference over `()` in this case.

**Comment by dbuenzli** [comment]
> invalid_arg "Format: unsychronized concurrent access"

As was established in this [PR](https://github.com/ocaml/ocaml/pull/11193#issuecomment-1298790120) there's no need for the "concurrent" but other than that I would much prefer if the stdlib fails loudly on race conditions rather than hiding them.

**Comment by craff** [comment]
Hello

Invalid_argument why not, but this is not really an invalid argument. A dedicated exception could be better than invalid_argument? Assert could be good to pointing to the code in format? 

Le 27 octobre 2024 22:40:13 GMT-10:00, Gabriel Scherer ***@***.***> a écrit :

**Comment by dbuenzli** [comment]
> Invalid_argument why not, but this is not really an invalid argument.

I think it's ok: you are giving an unsynchronized formatter to a formatting function. That's an invalid argument as per `Format`'s expectations.

> Assert could be good to pointing to the code in format?

I don't really care where the offending code is in `Format`. I only care to know that `Format`'s expectations were violated and that I'm responsible for it, which is what `Invalid_argument` is for. If I see an `Assert` failure I assume the bug is about an assumption internal to the library which is not satisfied (i.e. that's something I need to report to the library), which is not the case here.

**Comment by craff** [comment]
I updated the code to use invalid_arg and updated also the documentation to document this exception.

**Comment by craff** [comment]
I am wondering if we could make it systematic that this exception is raised in case of thread unsafe use of the library?

**Comment by gasche** [comment]
My understanding is that `invalid_arg` is used to signal a programming error (a function was called without respecting some documented precondition) that is not intended to be caught by the caller. I already use `Invalid_argument` in this way in the Dynarray module (see [documentation](https://ocaml.org/manual/5.2/api/Dynarray.html), the "Unsynchronized accesses" paragraph at the end of the toplevel comment), so in this sense it is arguably already a convention.

(Not completely fond of your error message but we are splitting hair. I'm sure @dbuenzli could make a good proposal and I would suggest to just reuse whatever he proposes.)

**Comment by dbuenzli** [comment]
perhaps something like: 
```
Format: Unsynchronized access to formatter 
```
or if the `Format` docs are good enough to allow the puzzled programmer to answer the question of what this could be: 
```
Format: Unsynchronized access to formatter, use a synchronized formatter
```

**Comment by craff** [comment]
> perhaps something like:
> 
> ```
> Format: Unsynchronized access to formatter 
> ```
> 

I just did that, and updated my modification mentionning the exception in Format.mli. 

I still wounder if we can capture a few more case of unsynchronised use.

Note: when I do logging in simple_httpd, I use one distinct file and formatter per domain. No mutex this way and 
the fist column being the unix epoch, it is easy to merge sort externaly.

File: Changes
Comment (gasche) [inline Changes]: and Daniel Bünzli

File: stdlib/format.ml
Comment (gasche) [inline stdlib/format.ml:447]: I'm not sure what the comment means, because we peeked into it *before*, but it suggests that we are now peeking. If I understand correctly, we have decided to commit, drop the element we saw when peeking, and process it. (If you're not sure what to write, I would maybe just drop the comment.)

File: stdlib/format.mli
Comment (gasche) [inline stdlib/format.mli]: I would not specify the payload exactly, so that we can tweak it later and still complain at users that would have tried to match on it literally. Just [Invalid_argument] should be enough information.

**Review by gasche — APPROVED** [review]
This looks fine to me (see minor comments below). I'm not sure whether this qualifies as a "stdlib change that requires a second approval", but in case I'd ask @Octachron or @nojb or @OlivierNicole for a second opinion.

File: stdlib/format.ml
Comment (craff) [reply stdlib/format.ml:447]: > I'm not sure what the comment means, because we peeked into it _before_, but it suggests that we are now peeking. If I understand correctly, we have decided to commit, drop the element we saw when peeking, and process it. (If you're not sure what to write, I would maybe just drop the comment.)

This comment was there before and I kept it unchanged. We peek_opt the element and if there is enough room we peek it and print it.

File: stdlib/format.mli
Comment (craff) [reply stdlib/format.mli]: Done

File: stdlib/format.ml
Comment (Octachron) [reply stdlib/format.ml:447]: The logic is rather: we peeked into the token queue, and if we know enough about the size of the token, we consume it. Otherwise, we wait for more information on this token size.

**Review by Octachron — APPROVED** [review]
Raising a clear exception for certain unsynchronized access sounds like a good to me too.

File: stdlib/format.mli
Comment (gasche) [inline stdlib/format.mli]: "undefined behavior" is understood as "this can segfault", so it is not the correct wording here. I would use: "This may result in [Invalid_argument] being raised, or in unspecified behaviors."

**Comment by gasche** [comment]
Merged (by squashing all commits), thanks!

**Comment by craff** [comment]
Thanks all for your comments!

File: Changes
Comment (SkySkimmer) [inline Changes:70]: incorrect number

**Comment by craff** [comment]
Gaëtan Gilbert ***@***.***> writes:

> @SkySkimmer commented on this pull request.
>
> -------------------------------------------------------------------------------------------------------------------------
>
> In Changes:
>
>> @@ -67,6 +67,10 @@ Working version
>
>  ### Standard library:
>
> +- #13371: Avoid Queue.empty to be raised by Format when used concurrently.

This is the number of the issue, not the number of the PR. I thought
this was OK too ?

> incorrect number
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.


--
Christophe Raffalli
web: https://raffalli.eu

**Comment by Octachron** [comment]
The convention is to cite all PRs and issues involved in the Changes entry (I will fix the Changes entry manually).

**Comment by craff** [comment]
> Gaëtan Gilbert ***@***.***> writes:
> @SkySkimmer commented on this pull request. ------------------------------------------------------------------------------------------------------------------------- In Changes: > @@ -67,6 +67,10 @@ Working version ### Standard library: +- #13371: Avoid Queue.empty to be raised by Format when used concurrently.
> This is the number of the issue, not the number of the PR. I thought this was OK too ?
> […](#)
> incorrect number — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.
> -- Christophe Raffalli web: https://raffalli.eu

Argh, no you are right, this is another issue with format. As this is merged, I do not know really what to do ?

**Comment by Octachron** [comment]
The  typo are now fixed.

**Comment by craff** [comment]
Florian Angeletti ***@***.***> writes:

Thanks and sorry for that typo and the time it took to you.

Cheers, Christophe.

> The typo are now fixed.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.


--
Christophe Raffalli
web: https://raffalli.eu

**Pull request merged**
